### PR TITLE
[XLA] Thunk to cudnn fp16 kernels for F16 batchnormalization and enable cudnn for batchnormalization by default

### PR DESCRIPTION
**Pull request by AyanmoI**
When using cuDNN for batchnormalization, xla currently thunks fp16 batchnorms to fp32 kernels by casting fp16 inputs and gradients to fp32 in the tf2xla bridge. This is non-optimal since cuDNN 7.6 has highly efficient `Half` kernels that can be leveraged. This PR modifies `cudnn_batchnorm_rewriter` and `cudnn_batchnorm_thunk` such that the converts introduced in the bridge get eliminated and the corresponding fp16 cudnn APIs get called. 
Note that instead of eliminating convert ops, I add converts which then get eliminated by `algebraic_simplifier`. [PR](https://github.com/tensorflow/tensorflow/pull/32435) adding capability to eliminate of convert pairs was merged earlier.

While investigating performance of mobilenet, it was found that enabling cuDNN for Batchnorm improves performance by around 10-12%. Running the numbers for a set of convolutional networks in addition to mobilenet with cuDNN turned on for Batchnorm, showed that enabling cuDNN, either improves performance or keeps it the same. This PR enables cudnn for batchnorm by default  based on the above-mentioned performance enhancements.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc [hunk 1]
```diff
@@ -195,15 +265,57 @@ Status Visitor::HandleBatchNormGrad(HloInstruction* batch_norm) {
 
   std::vector<HloInstruction*> operands(batch_norm->operands().begin(),
                                         batch_norm->operands().end());
+  bool is_convert_inserted = false;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Would it be too much work to split out `HandleBatchNormGrad` into a separate PR?  That way we'll be able to isolate performance issues more easily.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc [hunk 2]
```diff
@@ -195,15 +265,57 @@ Status Visitor::HandleBatchNormGrad(HloInstruction* batch_norm) {
...
+      if ((batch_norm->operand(index)->opcode() == HloOpcode::kConvert) &&
+          (batch_norm->operand(index)->operand(0)->shape().element_type() ==
+           F16) &&
+          (ShapeUtil::Compatible(
+              operand_0_convert->shape(),
+              batch_norm->mutable_operand(index)->shape()))) {
+        operands[index] = add_convert(batch_norm->mutable_operand(index), F16);
+      }
+    }
+    is_convert_inserted = true;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Same comment as above -- you probably want to call this `is_batchnorm_with_fp16_inputs` or something like that.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc [hunk 3]
```diff
@@ -109,30 +150,71 @@ Status CudnnBatchNormForwardInferenceThunk::ExecuteOnStream(
...
-      /*workspace_allocator=*/nullptr);
+  se::DeviceMemoryBase output_base =
+      buffer_allocations.GetDeviceAddress(output_);
+  PrimitiveType output_primitive_type =
+      hlo_instruction()->shape().element_type();
+
+  switch (output_primitive_type) {
+    case F16: {
+      auto output = se::DeviceMemory<Eigen::half>(output_base);
+      stream.ThenBatchNormalizationForward(
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc]: Extract out into a template function?

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc [hunk 4]
```diff
@@ -30,6 +30,47 @@ namespace gpu {
 
 namespace dnn = se::dnn;
 
+namespace {
+void IsInputOutputDtypeValid(const HloInstruction* hlo) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc]: - Since the function is imperative it should be called `CheckXYZ` not `IsXYZ`.
 - Dtype is a TF-ism.  In XLA it is called "primitive type".

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Comment by AyanmoI** [comment]
I'll create 2 separate PRs and address the comments.

**Comment by AyanmoI** [comment]
Pushed another commit incorporating most of the comments. Figured it would be convenient to address most of the comments here before migrating to the 2-PR submission. Will address the refactoring to use templates in the subsequent PR.

**Comment by sanjoy** [comment]
> Pushed another commit incorporating most of the comments. Figured it would be convenient to address most of the comments here before migrating to the 2-PR submission. Will address the refactoring to use templates in the subsequent PR.

Thanks Ayan, LMK when this is ready to review.

Btw, do you have a microbenchmark showing the performance difference between XLA's batchnorm and cudnn's batchnorm?

**Comment by AyanmoI** [comment]
Sorry for the delayed response. Using `tf_cnn_benchmarks` in https://github.com/tensorflow/benchmarks for mobilenet_v2, the most recent numbers that I have are as follows: 
With cudnn BN: ~4500 images/sec
With Fused kernel BN: ~3200 images/sec
This is on a GV100. Note that this is based on fairly new repo but not the latest. The latest tensorflow has some refactors that causes the benchmark script to fail (tensorflow contrib has been refactored somehow). The benchmark scripts haven't been updated to reflect those changes. A few weeks back, I had the numbers looking like this:
With cudnn BN: ~4500 images/sec
With Fused kernel BN: ~3800 images/sec
The numbers seems dynamic but it is clear that at this point cudnn batchnorm performs better than the kernel fusion. 
I actually investigated the kernel fusion for batchnorm and found that the difference is predominantly due to `GTE` thwarting fusion of multioutput fused kernels. Infact there are other cases where see that FusionFusion merger (i.e, fusion across `GTE` ) is the bottleneck.

**Comment by AyanmoI** [comment]
@sanjoy I have pushed the PR to thunk to cudnn (https://github.com/tensorflow/tensorflow/pull/32887). Please take a look. I will create a new PR with the `cudnn_batchnorm_rewriter` changes once this PR is merged. The rewriter changes are dependent on https://github.com/tensorflow/tensorflow/pull/32887 getting merged.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Will submit separate PR once https://github.com/tensorflow/tensorflow/pull/32887 gets merged.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc (hunk 3 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc]: Done in https://github.com/tensorflow/tensorflow/pull/32887

**Comment by gbaned** [comment]
@AyanmoI Can you please resolve conflicts? Thanks!

**Comment by AyanmoI** [comment]
@sanjoy Please take a look at https://github.com/tensorflow/tensorflow/pull/33259 which handles fp16 batchnorms in `cudnn_batchnorm_rewriter`. Please note that https://github.com/tensorflow/tensorflow/pull/33259  addresses all the comments you made in this PR.

**Comment by AyanmoI** [comment]
> @AyanmoI Can you please resolve conflicts? Thanks!

@gbaned This PR has been split into 2 parts actually. Part 1 https://github.com/tensorflow/tensorflow/pull/32887 has already been merged. Part 2 https://github.com/tensorflow/tensorflow/pull/33259 is up for review. I will close this PR eventually. I had just kept this around as a reference.
